price-estimator/src/filter.rs

```diff
@@ -256,29 +229,29 @@ async fn estimate_amounts_at_price<T>(
 }
 
 /// Like `estimate_amounts_at_price` but the price is given and returned in atoms.
-async fn estimate_amounts_at_price_atoms<T>(
+async fn estimate_amounts_at_price_atoms(
     token_pair: TokenPair,
     price_in_quote: f64,
-    price_rounding_buffer: f64,
-    orderbook: Arc<Orderbook<T>>,
+    orderbook: Arc<Orderbook>,
 ) -> EstimatedOrderResult {
-    // NOTE: The price in quote is `sell_amount / buy_amount` which is the
-    // inverse of an exchange rate. Additionally, we need to apply the price
-    // rounding buffer to the price, which will **increase** the exchange rate,
-    // making it more restrictive and the estimate more pessimistic.
-    let exchange_rate = 1.0 / apply_rounding_buffer(price_in_quote, price_rounding_buffer);
+    let exchange_rate = 1.0 / price_in_quote;
     let transitive_order = orderbook
-        .get_pricegraph()
+        .pricegraph()
         .await
         .order_at_exchange_rate(token_pair, exchange_rate);
     let (buy_amount_in_base, sell_amount_in_quote) = transitive_order
-        .map(|order| {
-            (
-                apply_rounding_buffer(order.buy, price_rounding_buffer),
-                order.sell,
-            )
-        })
+        .map(|order| (order.buy, order.sell))
         .unwrap_or_default();
+    // TODO: What to do about the rounding buffer? The order we get with this exchange rate
+    // is too optimistic because the solver will see it as if its sell amount had been reduced by
+    // the rounding buffer.
+    // If we increase the sell amount here to make up for this then buy_amount / sell_amount is no
+    // longer equal to price_in_quote.
+    // This might not be a problem in practice because the orderbook has already been
+    // conservatively worsened with double the rounding buffer than what the solver actually uses.
+    // Or we might be okay with the ratio equality no longer being exactly true.
```

Comment:
TODO

Reply:
If we change the buy amount by the same ratio then the result is no longer correct as in reality the solver will see a slightly lower sell amount which could mean that the order based on this won't get matched.

Reply:
I see. So in order to be completely correct we would have to pass the rounding buffer all the way down here: https://github.com/gnosis/dex-services/blob/04da5aa18a2bc790a21ac9d9b68a65f3911cd85e/pricegraph/src/api/price_estimation.rs#L93-L100

There we have the concrete amounts and could apply the rounding on both sides. But I'm not sure we want the rounding concept to be that intrusive.

We could also do an "iterative" approach where we apply the rounding buffer to the sell amount here in the top level method and then check via `estimate_buy_amount` if we can get a buyAmount at least as high as the desired ratio. Otherwise we need to reduce the amount until we do.

Reply:
I think we should be using the `order_for_limit_price` instead of `order_at_limit_price`. This will return true buy and sell amounts, such that `buy/sell > limit_price`. You can then use @fleupold suggestion and iterate to converge to proper volumes. In pseudo-code:
```rust
        let mut current_limit_price = limit_price;
        let mut order;
        while {
            order = pricegraph.order_for_limit_price(pair, current_limit_price);
            order.sell += rounding_buffer;
            order.limit_price() < limit_price
        } {
            current_limit_price = order.limit_price();
        }
        order.buy = order.sell * limit_price;
```